dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java

```diff
@@ -318,32 +370,327 @@ private List<Event> inlineAtomicEnd(FunctionCall ignored) {
+                // This thread would sleep here.  Explicit or spurious signals may wake it.
+                // Re-lock.
+                EventFactory.Pthread.newLock(lockAddress.toString(), lockAddress),
+                //TODO proper error code: ETIMEDOUT
+                EventFactory.newLocal(errorRegister, errorValue)
+        );
+    }
+
+    private List<Event> inlinePthreadKeyCreate(FunctionCall call) {
+        //see https://linux.die.net/man/3/pthread_key_create
+        final Register errorRegister = getResultRegisterAndCheckArguments(2, call);
+        final Expression key = call.getArguments().get(0);
+        final Expression destructor = call.getArguments().get(1);
+        final Program program = call.getFunction().getProgram();
+        final long threadCount = program.getThreads().size();
+        final int pointerBytes = types.getMemorySizeInBytes(types.getPointerType());
+        final Register object = call.getFunction().newRegister(types.getArchType());
+        final BigInteger sizeValue = BigInteger.valueOf((threadCount + 1) * pointerBytes);
+        final Expression size = expressions.makeValue(sizeValue, types.getArchType());
+        final BigInteger destructorOffsetValue = BigInteger.valueOf(threadCount * pointerBytes);
+        final Expression destructorOffset = expressions.makeValue(destructorOffsetValue, types.getArchType());
+        //TODO call destructor at each thread's normal exit
+        return List.of(
+                EventFactory.newAlloc(object, types.getArchType(), size, false),
+                EventFactory.newStore(key, object),
+                EventFactory.newStore(expressions.makeADD(object, destructorOffset), destructor),
+                assignSuccess(errorRegister)
+        );
+    }
+
+    private List<Event> inlinePthreadKeyDelete(FunctionCall call) {
+        //see https://linux.die.net/man/3/pthread_key_delete
+        final Register errorRegister = getResultRegisterAndCheckArguments(1, call);
+        //final Expression key = call.getArguments().get(0);
+        //final int threadID = call.getThread().getId();
+        //TODO the destructor should no longer be called by pthread_exit
+        return List.of(
+                assignSuccess(errorRegister)
+        );
+    }
+
+    private List<Event> inlinePthreadGetSpecific(FunctionCall call) {
+        //see https://linux.die.net/man/3/pthread_getspecific
+        final Register result = getResultRegisterAndCheckArguments(1, call);
+        final Register dummy = call.getFunction().newRegister(types.getArchType());
+        final Expression key = call.getArguments().get(0);
+        final int threadID = call.getThread().getId();
+        final Expression offset = expressions.makeValue(BigInteger.valueOf(threadID), types.getArchType());
+        return List.of(
+                EventFactory.newLoad(dummy, key),
+                EventFactory.newLoad(result, expressions.makeADD(dummy, offset))
+        );
```

Comment:
Are you sure the double load is correct? `set/getspecific` do not get a pointer as input but a value of type `pthread_key_t`. Because of that, a call to `getspecific(key)` should generate code like this;
```
keyVal = load(key); // Compiler-generated
call getspecific(keyVal);
```
And now you perform two more loads, which sounds wrong unless there is one more level of (unnecessary?) indirection. But your `pthread_create_key` implementation looks like `key` contains the pointer to the storage so there is no further level of indirection.